File: server/app/services/request_sort.py
"""Sorting + pagination for the authenticated DJ request list (issue #478).

Field sorts the database can express (date/upvotes/bpm/title/artist) run as
efficient ``ORDER BY ... LIMIT/OFFSET`` with a true ``COUNT``. Sorts needing
domain logic the DB can't express run in Python: ``key`` (harmonic Camelot
order) here, and ``best_match`` (priority scoring) in the endpoint. Every sort
ends with a deterministic ``id DESC`` tie-breaker so pages never duplicate or
skip rows as values reorder between polls.
"""

from datetime import datetime

from sqlalchemy import nullslast
from sqlalchemy.orm import Query, Session
from sqlalchemy.sql import func

from app.models.event import Event
from app.models.request import Request, RequestStatus
from app.schemas.request import RequestSort, SortDirection
from app.services.recommendation.camelot import parse_key

# Default direction per field when the client does not pass one explicitly.
DEFAULT_SORT_DIRECTION: dict[RequestSort, SortDirection] = {
    RequestSort.DATE_REQUESTED: SortDirection.DESC,
    RequestSort.DATE_ACCEPTED: SortDirection.DESC,
    RequestSort.UPVOTES: SortDirection.DESC,
    RequestSort.BPM: SortDirection.ASC,
    RequestSort.KEY: SortDirection.ASC,
    RequestSort.TITLE: SortDirection.ASC,
    RequestSort.ARTIST: SortDirection.ASC,
    RequestSort.BEST_MATCH: SortDirection.DESC,
}

# SQL columns for field sorts the database can order directly.
_SORT_COLUMNS = {
    RequestSort.DATE_REQUESTED: Request.created_at,
    RequestSort.DATE_ACCEPTED: Request.accepted_at,
    RequestSort.UPVOTES: Request.vote_count,
    RequestSort.BPM: Request.bpm,
    RequestSort.TITLE: func.lower(Request.song_title),
    RequestSort.ARTIST: func.lower(Request.artist),
}

# Sorts whose column is nullable — nulls always sort last, both directions.
_NULLABLE_SORTS = {RequestSort.DATE_ACCEPTED, RequestSort.BPM}


def filtered_requests_query(
    db: Session,
    event: Event,
    status: RequestStatus | None,
    since: datetime | None,
) -> Query:
    """Base query for an event's requests with the existing status/since filters."""
    query = db.query(Request).filter(Request.event_id == event.id)
    if status:
        query = query.filter(Request.status == status.value)
    if since:
        query = query.filter(Request.created_at > since)
    return query


def status_counts(db: Session, event: Event) -> dict[str, int]:
    """True per-status counts for an event, independent of pagination/filters.

    Backs the dashboard status-filter tabs (issue #478): a single
    ``GROUP BY status`` over the whole event, always returning all six keys
    (``all`` plus each status, 0 when absent). ``all`` is the cross-status total.
    """
    counts = {"all": 0, "new": 0, "accepted": 0, "playing": 0, "played": 0, "rejected": 0}
    rows = (
        db.query(Request.status, func.count(Request.id))
        .filter(Request.event_id == event.id)
        .group_by(Request.status)
        .all()
    )
    for status, count in rows:
        if status in counts:
            counts[status] = count
        counts["all"] += count
    return counts


def _camelot_ordinal(musical_key: str | None) -> int | None:
    """Map a key to a sortable harmonic ordinal (1A=2, 1B=3, ... 12B=25)."""
    pos = parse_key(musical_key)
    if pos is None:
        return None
    return pos.number * 2 + (1 if pos.letter == "B" else 0)


def key_sorted(rows: list[Request], direction: SortDirection) -> list[Request]:
    """Sort by harmonic key in Python; null/unparseable keys always last."""
    desc = direction == SortDirection.DESC

    def key_fn(r: Request) -> tuple:
        ordinal = _camelot_ordinal(r.musical_key)
        is_null = ordinal is None
        primary = (-ordinal if desc else ordinal) if ordinal is not None else 0
        return (
            is_null,
            primary,
            (r.song_title or "").lower(),
            (r.artist or "").lower(),
            -r.id,
        )

    return sorted(rows, key=key_fn)


def apply_field_sort(query: Query, sort: RequestSort, direction: SortDirection) -> Query:
    """Apply ``ORDER BY`` for a SQL-expressible field sort.

    Adds nulls-last for nullable columns and a deterministic ``id DESC``
    tie-breaker. Not valid for ``KEY`` (harmonic, Python-sorted via
    :func:`key_sorted`) or ``BEST_MATCH`` (priority-scored in the endpoint).
    """
    column = _SORT_COLUMNS[sort]
    ordering = column.asc() if direction == SortDirection.ASC else column.desc()
    if sort in _NULLABLE_SORTS:
        ordering = nullslast(ordering)
    return query.order_by(ordering, Request.id.desc())


def get_sorted_requests(
    db: Session,
    event: Event,
    *,
    status: RequestStatus | None,
    since: datetime | None,
    sort: RequestSort,
    direction: SortDirection,
    limit: int,
    offset: int,
) -> tuple[list[Request], int]:
    """Return ``(page_rows, true_total)`` for a field sort (not best_match)."""
    base = filtered_requests_query(db, event, status, since)
    total = base.count()

    if sort == RequestSort.KEY:
        ordered = key_sorted(base.all(), direction)
        return ordered[offset : offset + limit], total

    page = apply_field_sort(base, sort, direction).offset(offset).limit(limit).all()
    return page, total

File: dashboard/lib/request-sort.ts
/**
 * DJ request-list sort metadata (issue #478).
 *
 * The backend exposes 8 sort fields with per-field default directions. The
 * dashboard mirrors those defaults so that changing the sort field snaps the
 * direction toggle to the field's natural default; a separate toggle flips it.
 *
 * Keep `SORT_FIELDS` / `SORT_FIELD_DEFAULT_DIRECTION` in sync with
 * server/app/schemas (RequestSort + per-field defaults).
 */

import type { StatusFilter } from '@/app/(dj)/events/[code]/components/types';

import type { RequestSort, SortDirection } from './api-types';
import { camelotOrdinal } from './camelot';

/** Sort options in dashboard display order. Best Match leads (former "priority"). */
export const SORT_FIELDS: readonly { value: RequestSort; label: string }[] = [
  { value: 'best_match', label: 'Best Match' },
  { value: 'date_requested', label: 'Date requested' },
  { value: 'date_accepted', label: 'Date accepted' },
  { value: 'upvotes', label: 'Upvotes' },
  { value: 'bpm', label: 'BPM' },
  { value: 'key', label: 'Key' },
  { value: 'title', label: 'Title' },
  { value: 'artist', label: 'Artist' },
] as const;

/** Per-field default direction (mirrors the backend's per-field defaults). */
export const SORT_FIELD_DEFAULT_DIRECTION: Record<RequestSort, SortDirection> = {
  date_requested: 'desc',
  date_accepted: 'desc',
  upvotes: 'desc',
  bpm: 'asc',
  key: 'asc',
  title: 'asc',
  artist: 'asc',
  best_match: 'desc',
};

/** Default sort field when no preference is stored. */
export const DEFAULT_SORT_FIELD: RequestSort = 'date_requested';

/** Type guard: is this string one of the known sort fields? */
export function isRequestSort(value: string): value is RequestSort {
  return SORT_FIELDS.some((f) => f.value === value);
}

/**
 * Migrate a legacy `wrzdj-sort-${code}` value (the #468 toggle) to a sort field.
 * `'priority'` → `best_match`; anything else falls back to `date_requested`.
 */
export function migrateLegacySort(legacy: string | null): RequestSort {
  return legacy === 'priority' ? 'best_match' : DEFAULT_SORT_FIELD;
}

/**
 * Client-side sort/filter/count helpers (issue #489).
 *
 * The dashboard loads the whole bounded request set once, then sorts/filters/
 * counts in memory. These comparators MUST mirror
 * server/app/services/request_sort.py exactly so rows never jump between the
 * optimistic-patch view and the post-refetch view: every sort ends with a
 * deterministic `id DESC` tie-break; nullable `accepted_at`/`bpm` sort nulls-last
 * in both directions; `title`/`artist` compare case-insensitively; `key` sorts by
 * the Camelot ordinal with the same `(isNull, ordinal, title, artist, -id)` tuple
 * as `key_sorted`.
 */

/** Minimal structural row both SongRequest (main) and PendingReviewRow share. */
export interface SortableRequestRow {
  id: number;
  created_at: string;
  accepted_at?: string | null;
  vote_count: number;
  bpm?: number | null;
  musical_key?: string | null;
  song_title: string;
  artist: string;
  status: string;
}

/** The 7 fields sortable entirely client-side (everything except best_match). */
export type ClientSortField = Exclude<RequestSort, 'best_match'>;

const toTimestamp = (iso: string | null | undefined): number | null =>
  iso ? Date.parse(iso) : null;

/** Nulls-last numeric compare for a given direction (mirrors SQL nullslast). */
function compareNullableNumber(a: number | null, b: number | null, desc: boolean): number {
  const aNull = a === null;
  const bNull = b === null;
  if (aNull && bNull) return 0;
  if (aNull) return 1; // nulls last, both directions
  if (bNull) return -1;
  return desc ? b - a : a - b;
}

/** Case-insensitive string compare matching SQL `lower()` ordering (not locale). */
function compareString(a: string, b: string, desc: boolean): number {
  const x = a.toLowerCase();
  const y = b.toLowerCase();
  if (x === y) return 0;
  const cmp = x < y ? -1 : 1;
  return desc ? -cmp : cmp;
}

/** Pure comparator mirroring request_sort.py for a single client-sortable field. */
function comparator(field: ClientSortField, direction: SortDirection) {
  const desc = direction === 'desc';
  return (a: SortableRequestRow, b: SortableRequestRow): number => {
    let primary = 0;
    switch (field) {
      case 'date_requested':
        primary = compareNullableNumber(toTimestamp(a.created_at), toTimestamp(b.created_at), desc);
        break;
      case 'date_accepted':
        primary = compareNullableNumber(
          toTimestamp(a.accepted_at ?? null),
          toTimestamp(b.accepted_at ?? null),
          desc,
        );
        break;
      case 'upvotes':
        primary = desc ? b.vote_count - a.vote_count : a.vote_count - b.vote_count;
        break;
      case 'bpm':
        primary = compareNullableNumber(a.bpm ?? null, b.bpm ?? null, desc);
        break;
      case 'title':
        primary = compareString(a.song_title, b.song_title, desc);
        break;
      case 'artist':
        primary = compareString(a.artist, b.artist, desc);
        break;
      case 'key': {
        primary = compareNullableNumber(
          camelotOrdinal(a.musical_key),
          camelotOrdinal(b.musical_key),
          desc,
        );
        // key_sorted ties break by title then artist ASCENDING regardless of
        // direction, then by id DESC.
        if (primary === 0) primary = compareString(a.song_title, b.song_title, false);
        if (primary === 0) primary = compareString(a.artist, b.artist, false);
        break;
      }
    }
    return primary !== 0 ? primary : b.id - a.id; // id DESC tie-break
  };
}

/** Sort rows client-side for a simple field. Returns a new array (no mutation). */
export function sortRequests<T extends SortableRequestRow>(
  rows: readonly T[],
  field: ClientSortField,
  direction: SortDirection,
): T[] {
  return [...rows].sort(comparator(field, direction));
}

const EMPTY_STATUS_COUNTS: Record<StatusFilter, number> = {
  all: 0,
  new: 0,
  accepted: 0,
  playing: 0,
  played: 0,
  rejected: 0,
};

/** Per-status counts (+ all) from the in-memory set — correct by construction. */
export function computeStatusCounts(
  rows: readonly SortableRequestRow[],
): Record<StatusFilter, number> {
  const counts: Record<StatusFilter, number> = { ...EMPTY_STATUS_COUNTS };
  for (const r of rows) {
    counts.all += 1;
    // Own-property check (not `in`) so prototype keys like "toString" never match.
    if (r.status !== 'all' && Object.prototype.hasOwnProperty.call(counts, r.status)) {
      counts[r.status as StatusFilter] += 1;
    }
  }
  return counts;
}

/**
 * Coerce a server `status_counts` map into the strict StatusFilter-keyed record
 * (issue #521). Used on the capped path, where the in-memory set is truncated and
 * `computeStatusCounts` would undercount: the backend's authoritative counts are
 * pagination-independent. Missing keys default to 0; unknown keys are ignored.
 */
export function normalizeStatusCounts(
  raw: Record<string, number> | null | undefined,
): Record<StatusFilter, number> {
  const counts: Record<StatusFilter, number> = { ...EMPTY_STATUS_COUNTS };
  if (!raw) return counts;
  for (const key of Object.keys(counts) as StatusFilter[]) {
    const value = raw[key];
    if (typeof value === 'number' && Number.isFinite(value)) counts[key] = value;
  }
  return counts;
}

/** Filter rows by status tab; 'all' returns the full set. */
export function filterByStatus<T extends SortableRequestRow>(
  rows: readonly T[],
  filter: StatusFilter,
): T[] {
  if (filter === 'all') return [...rows];
  return rows.filter((r) => r.status === filter);
}
